src/o365/spo/commands/page/page-clientsidewebpart-add.ts

```diff
@@ -0,0 +1,483 @@
+
+interface CommandArgs {
+	options: Options;
+}
+
+interface Options extends GlobalOptions {
+	pageName: string;
+	webUrl: string;
+	standardWebPart: StandardWebPart;
+	webPartId: string;
+	webPartProperties: string;
+	section: number;
+	column: number;
+	order: number;
+}
+
+class SpoPageClientSideWebPartAddCommand extends SpoCommand {
+	public get name(): string {
+		return `${commands.PAGE_CLIENTSIDEWEBPART_ADD}`;
+	}
+
+	public get description(): string {
+		return 'Adds a client side WebPart to a modern page';
+	}
+
+	public commandAction(cmd: CommandInstance, args: CommandArgs, cb: (err?: any) => void): void {
+		const resource: string = Auth.getResourceFromUrl(args.options.webUrl);
+		let siteAccessToken: string = '';
+		let requestDigest: string = '';
+		let clientSidePage: ClientSidePage = null as any;
+
+		let pageName: string = args.options.pageName;
+		if (args.options.pageName.indexOf('.aspx') < 0) {
+			pageName += '.aspx';
+		}
+
+		if (this.debug) {
+			cmd.log(`Retrieving access token for ${resource}...`);
+		}
+
+		auth
+			.getAccessToken(resource, auth.service.refreshToken as string, cmd, this.debug)
+			.then((accessToken: string): request.RequestPromise => {
+				if (this.debug) {
+					cmd.log(`Retrieved access token ${accessToken}. Retrieving request digest...`);
+				}
+
+				siteAccessToken = accessToken;
+
+				if (this.verbose) {
+					cmd.log(`Retrieving request digest...`);
+				}
+
+				return this.getRequestDigestForSite(args.options.webUrl, siteAccessToken, cmd, this.debug);
+			})
+			.then((res: ContextInfo): Promise<ClientSidePage> => {
+				if (this.debug) {
+					cmd.log('Response:');
+					cmd.log(res);
+					cmd.log('');
+				}
+
+				// Keep the reference of request digest for subsequent requests
+				requestDigest = res.FormDigestValue;
+
+				if (this.verbose) {
+					cmd.log(`Retrieving Client Side Page ${pageName}...`);
+				}
+				// Get Client Side Page
+				return this._getClientSidePage(pageName, cmd, args, siteAccessToken);
+			})
+			.then((csPage) => {
+				if (this.debug) {
+					cmd.log(`Retrieved Client Side Page:`);
+					cmd.log(csPage);
+					cmd.log('');
+				}
+
+				// Keep the reference of client side page for subsequent requests
+				clientSidePage = csPage;
+
+				if (this.verbose) {
+					cmd.log(
+						`Retrieving Client Side definition for WebPart ${args.options.webPartId ||
+							args.options.standardWebPart}...`
+					);
+				}
+				// Get the WebPart according to arguments
+				return this._getWebPart(cmd, args, siteAccessToken);
+			})
+			.then((webPart) => {
+				if (this.debug) {
+					cmd.log(`Retrieved WebPart definition:`);
+					cmd.log(webPart);
+					cmd.log('');
+				}
+
+				if (this.verbose) {
+					cmd.log(`Setting Client Side WebPart layout and properties...`);
+				}
+				// Set the WebPart properties and layout (section, column and order)
+				this._setWebPartPropertiesAndLayout(clientSidePage, webPart, cmd, args);
```

Comment:
If this method is not returning a promise, why is the next block wrapped in a separate then clause?